sage_code/strong_uniform_bounds.py

```diff
@@ -0,0 +1,128 @@
+"""strong_uniform_bounds.py
+
+    Computes the strong uniform bounds from the second paper.
+
+    ====================================================================
+
+    This file is part of Isogeny Primes.
+
+    Copyright (C) 2023 Barinder S. Banwait and Maarten Derickx
+
+    Isogeny Primes is free software: you can redistribute it and/or
+    modify it under the terms of the GNU General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+    The authors can be reached at: barinder.s.banwait@gmail.com and
+    maarten@mderickx.nl.
+
+    ====================================================================
+
+"""
+
+from sage.all import Partitions, divisors, matrix, GF, lcm
+from itertools import product
+
+from .common_utils import get_weil_polys
+from .pil_integers import collapse_tuple
+
+def is_non_increasing(t):
+
+    for j in range(len(t)-1):
+        if t[j] < t[j+1]:
+            return False
+    return True
+
+
+def get_beta_mats_with_pow(F, pow=1):
+
+    frob_polys = get_weil_polys(F)
+    return [matrix.companion(f) ** pow for f in frob_polys]
+
+
+def remove_duplicates(list_of_e_tuples):
+
+    output = []
+
+    for t in list_of_e_tuples:
+        if is_non_increasing(t):
+            output.append(t)
+    return output
+
+def splitting_types(d):
+    """Implements step 1 of Algorithm 5.1, the possible splitting types
+
+    Args:
+        d (int): degree of number field
+    """
+
+    types = []
+
+    for r in range(1,d+1):
+        parts = list(Partitions(d, length=r))        
+        for a_part in parts:
+            divisors_of_partition = [divisors(t) for t in a_part]
+            possible_es = list(product(*divisors_of_partition))
+            possible_es = remove_duplicates(possible_es)
```

Comment:
I think we might be removing to much here. The integers occurring in partitions in sage are already sorted by size (largest first) so if the a_part = [3,2] is partition of 5 then [1,2] is a sequence of possible es that would be discarded so the splitting type {"es":[1,2], "fs"=[3,1]} would be thrown away. I guess your idea was that this is no problem since the equivalent splitting type {"es":[2,1], "fs"=[1,3]} would still occur, however [2, 3] is not in the output of  list(Partitions(5, length=2)) so this equivalent splitting type will not occur either, which is a serious problem.

Could you add d=5 as a testcase, when you fix this?

Speaking of which: I do not like the use of articles in variable names. They make things longer but often don't contribute to understandability.

Reply:
p.s. I think the easiest way to fix this is to also pass `a_part` to the  `remove_duplicates` and `is_non_increasing` functions and only do the check `t[j] < t[j+1]` if `a_part[i] = a_part[j]`.

Reply:
solution looks good!